lib/postgrex/notifications.ex

```diff
@@ -285,107 +329,194 @@ defmodule Postgrex.Notifications do
     end
   end
 
-  def handle_info(msg, s) do
-    %{protocol: protocol, listener_channels: channels, listeners: listeners} = s
-    opts = [notify: &notify_listeners(channels, listeners, &1, &2)]
+  def handle_info(msg, %{protocol: protocol, state: {mod, mod_state}} = state) do
+    opts = [notify: &mod.handle_notification(&1, &2, mod_state)]
 
     case Protocol.handle_info(msg, opts, protocol) do
       {:ok, protocol} ->
-        {:noreply, %{s | protocol: protocol}, s.idle_interval}
+        {:noreply, %{state | protocol: protocol}, state.idle_interval}
+
+      {:unknown, protocol} ->
+        maybe_handle(mod, :handle_info, [msg, mod_state], %{state | protocol: protocol})
 
       {error, reason, protocol} ->
-        reconnect_or_stop(error, reason, protocol, s)
+        reconnect_or_stop(error, reason, protocol, state)
     end
   end
 
-  defp reestablish_listener({ref, {channel, pid}}, {:ok, s, timeout}) do
-    case do_listen(channel, pid, ref, nil, s) do
-      {:noreply, s, _} -> {:cont, {:ok, s, timeout}}
-      error -> {:halt, error}
+  def handle_info(msg, %{state: {mod, mod_state}} = state) do
+    maybe_handle(mod, :handle_info, [msg, mod_state], state)
+  end
+
+  defp maybe_handle(mod, fun, args, state) do
+    if function_exported?(mod, fun, length(args)) do
+      handle(mod, fun, args, nil, state)
+    else
+      {:noreply, state}
     end
   end
 
-  defp do_listen(channel, pid, ref, from, s) do
-    s = update_in(s.listener_channels[channel], &((&1 || %{}) |> Map.put(ref, pid)))
+  defp handle(mod, fun, args, from, state) do
+    case apply(mod, fun, args) do
+      {:noreply, mod_state} ->
+        {:noreply, %{state | state: {mod, mod_state}}, state.idle_interval}
 
-    # If this is the first listener for the given channel,
-    # we need to actually issue the LISTEN query.
-    if map_size(s.listener_channels[channel]) == 1 do
-      listener_query("LISTEN \"#{channel}\"", {:ok, ref}, {:eventually, ref}, from, s)
-    else
-      if from, do: Connection.reply(from, {:ok, ref})
-      {:noreply, s, s.idle_interval}
+      {:query, query, mod_state} ->
+        opts = [notify: &mod.handle_notification(&1, &2, mod_state)]
+
+        state = %{state | state: {mod, mod_state}}
+
+        with {:ok, result, protocol} <- Protocol.handle_simple(query, opts, state.protocol),
+             {:ok, protocol} <- Protocol.checkin(protocol) do
+          state = %{state | protocol: protocol}
+
+          handle(mod, :handle_result, [result, mod_state], from, state)
+        else
+          {error, reason, protocol} ->
+            from && reply(from, {__MODULE__, reason})
+
+            {:noreply, state, _} = maybe_handle(mod, :handle_disconnect, [mod_state], state)
```

Comment:
I think we should move this to `reconnect_or_stop` and invoke it only if auto_reconnect is set to true?

Reply:
Hmm. I think it should run before any disconnect because it could be about more than changing the `mod_state`. For example, imagine you want some side-effect in the handler, e.g., notifying other processes that the connection is terminating.